lib/tty/key.c

```diff
@@ -1718,6 +1768,144 @@ get_key_code (int no_delay)
+         * any of complex sequence we are interested in.
+         */
+    complex_seq:
+        while (pending_keys != NULL && pending_keys[0] == ESC_CHAR && pending_keys[1] == '[')
+        {
+            tty_nodelay (TRUE);
+            while ((c = tty_lowlevel_getch ()) >= 0)
+                push_char (c);
+            tty_nodelay (FALSE);
+
+            seq_complex = pending_keys;
+            pending_keys = seq_append = NULL;
+
+            // Check for complex sequence until any KEY_* constant (above 8bit) or maximum length
+            for (i = 0; seq_complex[i] != '\0' && i < 32; i++)
+            {
+                if (seq_complex[i] > 0xFF)
+                {
+                    pending_keys = seq_append = &seq_complex[i];
+                    break;
+                }
+                seq_char_buffer[i] = seq_complex[i];
+            }
+            seq_char_buffer[i] = '\0';
+
+            const char *seq_char_buffer_ptr = &seq_char_buffer[2];
+            if (!parse_csi (&csi, &seq_char_buffer_ptr, &seq_char_buffer[i]))
+                continue;
+
+            char final_byte = *(seq_char_buffer_ptr - 1);
+            if (*seq_char_buffer_ptr != '\0')
+                pending_keys = seq_append = &seq_buffer[seq_char_buffer_ptr - seq_char_buffer];
+
+            /* Check for Kitty Keyboard Protocol including backward compatibility sequences:
+             *
+             * CSI parameters [u~ABCDEFHPQS]
+             *
+             * https://sw.kovidgoyal.net/kitty/keyboard-protocol/
+             *
+             * Throughout the process, handle possible following sequences, leaving them in the
+             * pending queue.
+             *
+             * Examples:
+             *
+             * \E[2~ or \E[2;1~ -> KEY_IC
+             * \E[1;2A -> KEY_M_SHIFT | KEY_UP
+             * \E[13;6u -> KEY_M_CTRL | KEY_M_SHIFT | KEY_ENTER
+             * \E[111;5u -> KEY_M_CTRL | 'o'
+             * \E[44:63;132u -> KEY_M_ALT | '?' (on Czech keyboard, with Num Lock on)
+             */
+
+            if ((final_byte == 'u' && csi.param_count >= 1)
+                || (final_byte == '~' && csi.param_count >= 1)
+                || (final_byte >= 'A' && final_byte <= 'F') || (final_byte == 'H')
+                || (final_byte >= 'P' && final_byte <= 'S'))
+            {
+                // Private mode is used for reporting protocol presence and mode. We don't need it
+                if (csi.private_mode != 0)
+                    continue;
+
+                // Retrieve the modifier bits
+                if (csi.param_count >= 2)
+                    modifiers = ((csi.params[1][0] - 1) << 12) & KEY_M_MASK;
+                else
+                    modifiers = 0;
+
+                // Zero parameters are allowed. In that case, assume default 1 as the default
+                if (csi.param_count == 0)
+                    csi.params[0][0] = 1;
+
+                // Check for non-alphabet keys using a predefined table
```

Comment:
i wonder whether a cleaner solution (that isn't susceptible to failure due to the key list growing) would be possible ...